File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
    bool ispossible(vector<int> &nums, long long mid, int threshold)
    {
        long long sum=0;
        for(int i=0;i<nums.size();i++)
        {
            if(nums[i]%mid==0)
            {
                sum+=(nums[i]/mid);
            }
            else
            {
                sum+=(nums[i]/mid)+1;
            }
        }
        if(sum<=threshold)
        {
            return true;
        }
        return false;
    }

public:
    int smallestDivisor(vector<int>& nums, int threshold) {
        long long ans=-1;
        long long s=1;
        long long maxi=INT_MIN;
        for(int i=0;i<nums.size();i++)
        {
            if(nums[i]>maxi)
            {
                maxi=nums[i];
            }
        }
        long long e=maxi;
        while(s<=e)
        {
            long long mid=s+(e-s)/2;
            if(ispossible(nums,mid,threshold))
            {
                ans=mid;
                e=mid-1;
            }
            else
            {
                s=mid+1;
            }
        }
        return ans;
    }
};

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.py
class Solution:
    def ispossible(self, nums: List[int], mid: int, threshold: int) -> int:
        summ=0
        for i in range(len(nums)):
            if(nums[i]%mid==0):
                summ+=(nums[i]//mid)
            else:
                summ+=(nums[i]//mid)+1
        if(summ<=threshold):
            return True
        return False

    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        ans=-1
        s=1
        maxi=float('-inf')
        for i in range(len(nums)):
            if(nums[i]>maxi):
                maxi=nums[i]
        e=maxi
        while(s<=e):
            mid=s+(e-s)//2
            if(self.ispossible(nums,mid,threshold)):
                ans=mid
                e=mid-1
            else:
                s=mid+1
        return ans